Closes #7123: Rules about get_subscribed_events

Miraeld opened the pull request:
# Description

Fixes #7123 

This won't impact users.

## Type of change

- [x] Enhancement (non-breaking change which improves an existing functionality).
- [x] Sub-task of #7122 

## Detailed scenario

Just run PHPStan

## Technical description

### Documentation

This pull request introduces several changes related to PHPStan integration and testing in the project. The most significant updates include adding new PHPStan rules and tests, modifying the configuration files, and ensuring callback methods exist in subscribed events.

### PHPStan Integration and Configuration:
* [`composer.json`](diffhunk://#diff-d2ab9925cad7eac58e0ff4cc0d251a937ecf49e4b6bf57f8b95aab76648a9d34R237-R238): Added new scripts to run PHPStan with specific options (`run-stan-reset-baseline`, `run-stan-test`).
* [`phpstan.neon.dist`](diffhunk://#diff-6f19df6a6307a48db0940e6897591fb08776d2db8a6134737aa708defdb5c92dR3-R16): Updated parameters to include new paths and rules, and modified reporting settings. [[1]](diffhunk://#diff-6f19df6a6307a48db0940e6897591fb08776d2db8a6134737aa708defdb5c92dR3-R16) [[2]](diffhunk://#diff-6f19df6a6307a48db0940e6897591fb08776d2db8a6134737aa708defdb5c92dR58-R59)

### New PHPStan Rule:
* [`tests/phpstan/Rules/EnsureCallbackMethodsExistsInSubscribedEvents.php`](diffhunk://#diff-a11739111a2dd639d556c619f12f45f865a31811fb9c3e6becb7f60e884c0406R1-R152): Implemented a new PHPStan rule to ensure callback methods exist in subscribed events.

### Tests for New Rule:
* [`tests/phpstan/tests/Rules/EnsureCallbackMethodsExistsInSubscribedEventsTest.php`](diffhunk://#diff-4013e141054338bdda841d49497f0cb638e537988adbd81d5c425120f60f4186R1-R45): Added tests to validate the new PHPStan rule, including tests for valid subscribers, non-existing methods, and complex syntax.
* Added supporting test data files for the new rule:
  * `tests/phpstan/tests/data/EnsureCallbackMethodsExistsInSubscribedEventsTest/complex-syntax.php`
  * `tests/phpstan/tests/data/EnsureCallbackMethodsExistsInSubscribedEventsTest/not-existing.php`
  * `tests/phpstan/tests/data/EnsureCallbackMethodsExistsInSubscribedEventsTest/valid.php`

### PHPUnit Configuration:
* [`tests/phpstan/tests/bootstrap.php`](diffhunk://#diff-49fb88d28df699b5707a934fea9b23df83a82dd59ef84fa0825c088747ca680eR1-R3): Added bootstrap file for PHPUnit tests.
* [`tests/phpstan/tests/phpunit.xml`](diffhunk://#diff-5048354aeca46b8554885e0820d189163b77ae2fa3e177a36cb9418bd7d476e2R1-R27): Added PHPUnit configuration file to define test suites and coverage settings.

### New dependencies

None

### Risks

None

# Mandatory Checklist

## Code validation

- [x] I validated all the Acceptance Criteria. If possible, provide screenshots or videos.
- [x] I triggered all changed lines of code at least once without new errors/warnings/notices.
- [x] I implemented built-in tests to cover the new/changed code.

## Code style
- [x] I wrote a self-explanatory code about what it does.
- [x] I protected entry points against unexpected inputs.
- [x] I did not introduce unnecessary complexity.
- [x] Output messages (errors, notices, logs) are explicit enough for users to understand the issue and are actionnable.

Miraeld committed:
Rule to check if method of subscribed events exists

CrochetFeve0251 commented on tests/phpstan/Rules/EnsureCallbackMethodsExistsInSubscribedEvents.php:
```diff
@@ -0,0 +1,108 @@
+
+			// Get the function/method name from the scope
+			$functionName = $scope->getFunctionName();
+			if ('get_subscribed_events' === $functionName) {
```
@Miraeld two things:
- It would be better to bail out here to keep the logic simple.
- We also need to check that we extend from the interface `Subscriber_Interface` to prevent side effects

CrochetFeve0251 commented on tests/phpstan/Rules/EnsureCallbackMethodsExistsInSubscribedEvents.php:
```diff
@@ -0,0 +1,108 @@
+			$functionName = $scope->getFunctionName();
+			if ('get_subscribed_events' === $functionName) {
+				// Check if the return expression is an array
+				if ($node->expr instanceof Node\Expr\Array_) {
```
We need to bail out here too

CrochetFeve0251 commented on tests/phpstan/Rules/EnsureCallbackMethodsExistsInSubscribedEvents.php:
```diff
@@ -0,0 +1,108 @@
+				// Check if the return expression is an array
+				if ($node->expr instanceof Node\Expr\Array_) {
+					foreach ($node->expr->items as $item) {
+						if ($item instanceof Node\Expr\ArrayItem) {
```
Same here with a `continue`

CrochetFeve0251 commented on tests/phpstan/Rules/EnsureCallbackMethodsExistsInSubscribedEvents.php:
```diff
@@ -0,0 +1,108 @@
+
+		if ($methodValue instanceof Node\Scalar\String_) {
+			// Simple structure: array('hook_name' => 'method_name')
+			$errors = $this->checkIfMethodExistsInClass($methodValue->value, $scope, $methodValue);
```
We can bail out here

CrochetFeve0251 commented on tests/phpstan/Rules/EnsureCallbackMethodsExistsInSubscribedEvents.php:
```diff
@@ -0,0 +1,108 @@
+		} elseif ($methodValue instanceof Node\Expr\Array_) {
+			// More complex structures: array or nested array
+			foreach ($methodValue->items as $subItem) {
+				if ($subItem instanceof Node\Expr\ArrayItem) {
```
Same here, we can bail out

CrochetFeve0251 commented on tests/phpstan/Rules/EnsureCallbackMethodsExistsInSubscribedEvents.php:
```diff
@@ -0,0 +1,108 @@
+			// More complex structures: array or nested array
+			foreach ($methodValue->items as $subItem) {
+				if ($subItem instanceof Node\Expr\ArrayItem) {
+					if ($subItem->value instanceof Node\Scalar\String_) {
```
and same here with `continue`

CrochetFeve0251 committed:
Added a test to validate the logic

Miraeld committed:
Modify rule to work properly

Miraeld commented:
I've treated your feedback @CrochetFeve0251 , 
Also, modified the way it was working a bit to comply with your tests. 

Also, you'll notice I've added some `// @php-stan-ignore-[next]-line` 
That's because when phpstan run within this class, it kinda gets messed up with some type, even tho it's explicit, and due to this it thinks some part of the code are unreachable while in real they are. 

I add a screenshot of your test passing as it isn't part of the CI in here. 

![Screenshot 2024-11-25 at 02 44 29](https://github.com/user-attachments/assets/4226ccb1-69d7-4f39-a241-1b244298840b)

CrochetFeve0251 commented on tests/phpstan/Rules/EnsureCallbackMethodsExistsInSubscribedEvents.php:
```diff
@@ -0,0 +1,152 @@
+
+		return [
+			RuleErrorBuilder::message( $error_message )
+				->line( $node->getLine() ) // Add the line number.
```
@Miraeld can you add an identifier?

Miraeld replied:
Done `callbackMethodNotFound` has been added

wordpressfan committed:
Merge branch 'develop' into enhancement/7123-add-rule-for-get-subscribed-events

# Conflicts:
#	composer.json
#	phpstan.neon.dist

Miraeld committed:
Add usage of baseline

Miraeld committed:
Add identifier to error
